osgar/drivers/vesc.py

```diff
@@ -0,0 +1,144 @@
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+        self.debug_arr = []
+        self.prev = {}
+        self.verbose = False  # TODO move into Node
+        self.packet = []
+        self.tachometer = None
+
+    def update(self):
+        channel = super().update()
+        assert channel == 'can', channel
+        msg_id, data, flags = self.can  # via update()
+        if (msg_id & 0xF00) == 0x500:
+            assert data[0] % 7 == 0, data[0]
+            self.packet.extend(data[1:])
+            # the structure for version 3.40 has 59 bytes and looks like:
+            # https://github.com/vedderb/bldc/blob/43c3bbaf91f5052a35b75c2ff17b5fe99fad94d1/commands.c#L166-L185
+            """
+case COMM_GET_VALUES:
+ind = 0;
+send_buffer[ind + +] = COMM_GET_VALUES;
+buffer_append_float16(send_buffer, mc_interface_temp_fet_filtered(), 1e1, &ind);
+buffer_append_float16(send_buffer, mc_interface_temp_motor_filtered(), 1e1, &ind);
+buffer_append_float32(send_buffer, mc_interface_read_reset_avg_motor_current(), 1e2, &ind);
+buffer_append_float32(send_buffer, mc_interface_read_reset_avg_input_current(), 1e2, &ind);
+buffer_append_float32(send_buffer, mc_interface_read_reset_avg_id(), 1e2, &ind);
+buffer_append_float32(send_buffer, mc_interface_read_reset_avg_iq(), 1e2, &ind);
+buffer_append_float16(send_buffer, mc_interface_get_duty_cycle_now(), 1e3, &ind);
+buffer_append_float32(send_buffer, mc_interface_get_rpm(), 1e0, &ind);
+buffer_append_float16(send_buffer, GET_INPUT_VOLTAGE(), 1e1, &ind);
+buffer_append_float32(send_buffer, mc_interface_get_amp_hours(false), 1e4, &ind);
+buffer_append_float32(send_buffer, mc_interface_get_amp_hours_charged(false), 1e4, &ind);
+buffer_append_float32(send_buffer, mc_interface_get_watt_hours(false), 1e4, &ind);
+buffer_append_float32(send_buffer, mc_interface_get_watt_hours_charged(false), 1e4, &ind);
+buffer_append_int32(send_buffer, mc_interface_get_tachometer_value(false), &ind);
+buffer_append_int32(send_buffer, mc_interface_get_tachometer_abs_value(false), &ind);
+send_buffer[ind + +] = mc_interface_get_fault();
+buffer_append_float32(send_buffer, mc_interface_get_pid_pos_now(), 1e6, &ind);
+send_buffer[ind + +] = app_get_configuration()->controller_id;
+commands_send_packet(send_buffer, ind);
+"""
+            # the "tachometer" data are stored as 32bit signed integer starting
+            # from 45th byte. Because the data blocks contain 7 bytes we are
+            # interested in block with "address pointer" set to 42.
+            # Then we do not have to re-contruct all four independent bufferes
+            # but only "pick" the interesting sub-packet. Yes, there is a danger
+            # that checksum of the whole buffer was not valid!
+            if data[0] == 42:  # optimization - pick only this packet part
+                b = bytes(data[1:])
+                prev = self.tachometer
+                self.tachometer = struct.unpack_from('>i', b, 3)[0]
```

Comment:
Looking at the structure, the tacho starts at byte 44. If the buffer is sent using several CAN messages and the first byte says offset - which is 42 in this case - then we need to set the offset to 2 and not 3. Is my reasoning correct?

Reply:
I added offsets into structure description ... maybe `id + +` was misleading as it was copy and paste from robotika.cz where ++ is interpreted as new code. So 42 and 45 is correct in my opinion.

Reply:
I uploaded some sample data from 1m backward motion:
http://osgar.robotika.cz/subt/kloubak/vesc/test-vesc-190730_183238.log
`python -m osgar.replay --module vesc --draw test-vesc-190730_183238.log`